app/com/linkedin/drelephant/ElephantRunner.java

```diff
@@ -45,136 +48,377 @@
+            } else {
+              // Delete and save to prevent Optimistic Locking
+              jobInDb.delete();
+              result.save();
+            }
+          } else {
+            // Should not reach here. We consider only SUCCEEDED and FAILED JOBs in _completedJobQueue
+            throw new RuntimeException(analyticJob.getAppId() + " is in " + jobState.name() + " state. This should not"
+                + " happen. We consider only Succeeded and Failed jobs in Completed Jobs Queue");
+          }
         } catch (InterruptedException ex) {
           Thread.currentThread().interrupt();
         } catch (Exception e) {
           logger.error(e.getMessage());
           logger.error(ExceptionUtils.getStackTrace(e));
+          retryAndDrop(analyticJob);
+        }
+      }
+      logger.info("Executor Thread" + _threadId + " for completed jobs is terminated.");
+    }
+  }
 
-          if (analyticJob != null && analyticJob.retry()) {
-            logger.error("Add analytic job id [" + analyticJob.getAppId() + "] into the retry list.");
-            _analyticJobGenerator.addIntoRetries(analyticJob);
-          } else {
-            if (analyticJob != null) {
-              logger.error("Drop the analytic job. Reason: reached the max retries for application id = ["
-                      + analyticJob.getAppId() + "].");
+  private class UndefinedExecutorThread implements Runnable {
+
+    private int _threadId;
+    private DelayQueue<AnalyticJob> _undefinedJobQueue;
+
+    UndefinedExecutorThread(int threadNum, DelayQueue<AnalyticJob> undefinedJobQueue) {
+      this._threadId = threadNum;
+      this._undefinedJobQueue = undefinedJobQueue;
+    }
+
+    @Override
+    public void run() {
+      while (_running.get() && !Thread.currentThread().isInterrupted()) {
+        AnalyticJob analyticJob = null;
+        try {
+          analyticJob =_undefinedJobQueue.take();
+          Job job = getJobFromCluster(analyticJob);
+          analyticJob.setJobStatus(job.getJobState().name()).setSeverity(0);
+          JobStatus.State jobState = job.getJobState();
+
+          // Job State can be RUNNING, SUCCEEDED, FAILED, PREP or KILLED
+          if (jobState == JobStatus.State.SUCCEEDED || jobState == JobStatus.State.FAILED) {
+            logger.info("App " + analyticJob.getAppId() + " has now completed. Adding it to the completed job Queue.");
+            delayedEnqueue(analyticJob.setFinishTime(job.getFinishTime()), _completedJobQueue);
+          } else if (jobState == JobStatus.State.RUNNING) {
+            logger.info("Executor thread " + _threadId + " for undefined jobs is analyzing "
+                + analyticJob.getAppType().getName() + " " + analyticJob.getAppId() + " :: " + jobState.name());
+            if (analyticJob.getAppType().getName().equalsIgnoreCase(SPARK_APP_TYPE)) {
+              // Ignore as we do not capture any metrics for Spark jobs now
+              delayedEnqueue(analyticJob, _undefinedJobQueue);
```

Comment:
We are now fetching jobs by a start time range. So when we fetch a spark job which is running, it moves to the runningJobQueue. We can move this to the completedJobQueue only after it has completed. So we check the status of the spark job after a update interval and retain it in the undefinedJobQueue as long as it is not completed.